### DISCUSSION TITLE
How to integrate Paypal OAuth in Pocketbase

### DISCUSSION THREAD

**Discussion by shubhendumadhukar** [pocketbase/pocketbase#4622] [Q&A] [answered]
Since PayPal is not in the list of supported auth providers, is there a way I can still configure Pocketbase to support PayPal auth? I see something called OpenID Connect, wondering if that could be used. I am new to OAuth, so this is an unchartered territory for me.

My specific requirement is to allow users to connect their PayPal to their existing user account in pocketbase.

**Comment by ganigeorgiev** [comment, marked as answer]
I haven't tried it but from a quick look at the PayPal documentation I guess it should be possible to use the existing PocketBase OIDC client.

First you need create the OAuth2 app on the PayPal side (see ["Enable Log in with PayPal"](https://developer.paypal.com/docs/log-in-with-paypal/integrate/#link-enableloginwithpaypal)). For the redirect url you'll have to specify `http://yourapp.com/api/oauth2-redirect` (_in case of the "all-in-one" flow_) or your own custom redirect handler (_if you use the "Manual code exchange" flow_).

After obtaining client id and secret, based on the PayPal [`.well-known` json](https://www.paypal.com/.well-known/openid-configuration), the OIDC endpoints that you need to set in the PocketBase Admin UI are:
- Auth url: `https://www.paypal.com/signin/authorize`
- Token url: `https://api.paypal.com/v1/oauth2/token`
- User API url: `https://api.paypal.com/v1/oauth2/token/userinfo`

After that you should be able to authenticate with the provider by calling something like:

```js
// for more details, please check https://pocketbase.io/docs/authentication/#oauth2-integration
await pb.collection("users").authWithOAuth2({ provider: "oidc" })
```

**Reply by shubhendumadhukar** [reply]
It almost worked. At least the implementation from Pocketbase side is working as expected. Some issue with PayPal endpoints. The userinfo endpoint throws a 400 error. I will look into that.

Thank you!

**Reply by ganigeorgiev** [reply]
> The userinfo endpoint throws a 400 error.

It is possible that is failing your `users` collection Create API rule or field constraints (eg. nonempty).
You can check for a more detailed error message in the _Admin UI > Logs_ and/or start the executable with `--dev`.

**Reply by shubhendumadhukar** [reply]
Did check the logs, which is why I assumed it's something to do with paypal.

```
failed to fetch OAuth2 user profile via https://api.paypal.com/v1/oauth2/token/userinfo (400):
{"error":"invalid_request","error_description":"Unable to fetch details to serve the request"}
```

My create rule is empty.

**Comment by ArthurRM** [comment]
> Did check the logs, which is why I assumed it's something to do with paypal.
> 
> ```
> failed to fetch OAuth2 user profile via https://api.paypal.com/v1/oauth2/token/userinfo (400):
> {"error":"invalid_request","error_description":"Unable to fetch details to serve the request"}
> ```
> 
> My create rule is empty.

Did you ever manage to fix this?
